File: MusicKit/Applications/PianoRoll/PartView.m
/* $Id$ */

#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>
#import <MusicKitLegacy/MusicKitLegacy.h>
#import <math.h>
#import "PartView.h"
#import "TadPole.h"

#define DEFAULT_BEATSCALE 32
#define DEFAULT_FREQSCALE 16

@implementation PartView

/* used to quickly identify PartView in a NSWindow */
- (NSInteger) tag
{
    return 1;
}

- setScore: (MKScore *) aScore
{
    NSArray *theParts;
    MKPart *thePart;
    MKNote *theNote;
    TadPole *newTad;
    int scoreDuration;
    NSInteger i, j, k, partCount;
    NSRect aRect;
    
    beatScale = DEFAULT_BEATSCALE;
    freqScale = DEFAULT_FREQSCALE;
    selectedList = [[NSMutableArray alloc] init];
    
    theParts = [aScore parts];
    partCount = [theParts count];
    
    scoreDuration = 0;			/* find the length of the score */
    for (i = 0; i < [theParts count]; i++) {
        thePart = [theParts objectAtIndex:i];
        for (j = 0; j < [thePart countOfNotes]; j++) {
            theNote = [thePart noteAtIndex:j];
            if ([theNote noteType] == MK_noteDur) {
                if ([theNote timeTag] + [theNote duration] > scoreDuration)
                    scoreDuration = [theNote timeTag] + [theNote duration];
            }
            else if ([theNote timeTag] > scoreDuration)
                scoreDuration = [theNote timeTag];
        }
    }
    if(scoreDuration * beatScale > [self bounds].size.width) {  
        beatScale = [self bounds].size.width / scoreDuration;
    }
    aRect = NSMakeRect(0.0, 0.0, scoreDuration*beatScale, log(MAXFREQ)*freqScale);
//    [self initWithFrame:aRect];

    for (i = 0; i < partCount; i++) {
        thePart = [theParts objectAtIndex:i];
        for (j = 0; j < [thePart countOfNotes]; j++) {
            theNote = [thePart noteAtIndex: j];
            switch ([theNote noteType]) {
            case MK_mute:
            case MK_noteOff:
                break;
            case MK_noteDur:
                newTad = [[TadPole alloc] initNote: theNote
                                            second: nil
                                           partNum: i
                                         beatscale: beatScale
                                         freqscale: freqScale];
                [self addSubview:newTad];
                break;
            case MK_noteUpdate:
                if ([theNote noteTag] == MAXINT)  /* no note tag! */
                    break;
            case MK_noteOn:
                if (MKIsNoDVal([theNote frequency])) /* no frequency - not quite kosher */
                    break;
                for (k = j+1; k < [thePart countOfNotes]; k++)
                    if ([[thePart noteAtIndex:k] noteTag] == [theNote noteTag])
                        break;
                if (k < [thePart countOfNotes]) {
                    newTad = [[TadPole alloc] initNote:theNote
                                                second:[thePart noteAtIndex:k]
                                                partNum:i
                                                beatscale:beatScale
                                                freqscale:freqScale];
                    [self addSubview:newTad];
                    break;
                }
            }
        }
    }
    return self;
}

- (void)gotClicked:sender with:(NSEvent *)theEvent
{
	BOOL loop = YES, inside, wasdragged = NO, wasSelected = YES;
	NSPoint ePoint, thePoint;
	NSEvent *nEvent;
        int i;
	id theTad;
		
	if (![sender isSelected]) {
		wasSelected = NO;
		[sender setMoving:YES];
		[sender doHighlight];
		for (i = 0; i < [selectedList count]; i++)
                [[selectedList objectAtIndex:i] unHighlight];
                [selectedList removeAllObjects];
		[selectedList addObject:sender];
	}
		else [sender erase];
//	ePoint.x = theEvent->location.x;
//	ePoint.y = theEvent->location.y;
//	ePoint = [self convertPoint:ePoint fromView:nil];
        ePoint = [theEvent locationInWindow];
	while (loop) {
		nEvent = [[self window] nextEventMatchingMask:(NSEventMaskLeftMouseUp |			NSEventMaskLeftMouseDragged)];
		thePoint.x = [nEvent locationInWindow].x;
		thePoint.y = [nEvent locationInWindow].y;
		thePoint = [self convertPoint:thePoint fromView:nil];
		inside = NSMouseInRect(thePoint , [self bounds] , [self isFlipped]);
		switch ([nEvent type]) {
		case NSEventTypeLeftMouseUp:
			loop = NO;
			if (!wasdragged && wasSelected) {
				[sender unHighlight];
				[selectedList removeObject:sender];
			}
			else {
				for (i = 0; i < [selectedList count]; i++) {
                                theTad = [selectedList objectAtIndex:i];
					[theTad setFromPosWith:beatScale :freqScale];
					[theTad setMoving:NO];
					[theTad display];
				}
			}
			break;
		case NSEventTypeLeftMouseDragged:
			wasdragged = YES;
			for (i = 0; i < [selectedList count]; i++) {
                        theTad = [selectedList objectAtIndex:i];
				[theTad setMoving:YES];
			//	[theTad moveBy:thePoint.x - ePoint.x :thePoint.y - ePoint.y];
                                [theTad setFrameOrigin:NSMakePoint(([theTad frame].origin.x + thePoint.x - ePoint.x), ([theTad frame].origin.y + thePoint.y - ePoint.y))];
				[theTad display];
			}
			ePoint.x = thePoint.x;
			ePoint.y = thePoint.y;
			break;
                default:        // Makes explicit to the compiler we do nothing with other events.
                        break;
		}
	}
}

- (double) beatScale
{
	return beatScale;
}

- (double)freqScale
{
	return freqScale;
}

- (void)setBeatScale:(double)bscale
{
	beatScale = bscale; 
}

- (void)setFreqScale:(double)fscale
{
	freqScale = fscale; 
}

- (void) documentView
{
}

@end

File: MusicKit/Applications/PianoRollSwift/PartView.swift
//
//  PlayScore.swift
//  PianoRollSwift
//
//  Created by C.W. Betts on 1/31/23.
//  Copyright © 2023 MusicKit Project. All rights reserved.
//

import Cocoa
import MusicKitLegacy
import MusicKitLegacy.MKPlugin

let MAXFREQ: Double = 7040

private let DEFAULT_BEATSCALE: Double = 32
private let DEFAULT_FREQSCALE: Double = 16

class PartView: NSView {
	
	var beatScale: Double = DEFAULT_BEATSCALE
	var frequencyScale: Double = DEFAULT_FREQSCALE
	private var selectedList = [TadPole]()

	/* used to quickly identify PartView in a NSWindow */
	override var tag: Int {
		return 1
	}
	
	func setScore(_ aScore: MKScore) {
		beatScale = DEFAULT_BEATSCALE
		frequencyScale = DEFAULT_FREQSCALE

		let theParts = aScore.parts
		let partCount = theParts.count

		var scoreDuration: Double = 0
		/* find the length of the score */
		for thePart in theParts {
			for j in 0 ..< thePart.countOfNotes {
				let theNote = thePart.note(at: j)!
				if theNote.noteType == .noteDur {
					if theNote.timeTag + theNote.duration() > scoreDuration {
						scoreDuration = theNote.timeTag + theNote.duration()
					}
				} else if theNote.timeTag > scoreDuration {
					scoreDuration = theNote.timeTag
				}
			}
		}
		if scoreDuration * beatScale > bounds.size.width {
			beatScale = bounds.size.width / scoreDuration
		}
		frame = NSRect(origin: frame.origin, size: CGSize(width: scoreDuration*beatScale, height: log(MAXFREQ)*frequencyScale))
		
		for (i, thePart) in theParts.enumerated() {
			for j in 0 ..< thePart.countOfNotes {
				let theNote = thePart.note(at: j)!
				switch theNote.noteType {
				case .noteDur:
					let newTad = TadPole(note: theNote, second: nil, partNumber: i, beatScale: beatScale, frequencyScale: frequencyScale)
					addSubview(newTad)
				case .noteUpdate:
					if theNote.noteTag == .max { /* no note tag! */
						break
					}
					fallthrough
				case .noteOn:
					if MKIsNoDVal(theNote.frequency) { /* no frequency - not quite kosher */
						break;
					}
					for k in (j + 1) ..< thePart.countOfNotes {
						let newTad = TadPole(note: theNote, second: thePart.note(at: k), partNumber: i, beatScale: beatScale, frequencyScale: frequencyScale)
						addSubview(newTad)
					}
				case .noteOff, .mute:
					break;
					
				@unknown default:
					break
				}
				
			}
		}
	}
	
	func gotClicked(by sender: TadPole, with theEvent: NSEvent) {
		var loop = true, inside = true, wasdragged = false, wasSelected = true;
			
		if (!sender.isSelected) {
			wasSelected = false
			sender.isMoving = true
			sender.highlight()
			for selected in selectedList {
				selected.unhilight()
			}
			selectedList.removeAll()
			selectedList.append(sender)
		} else {
			sender.erase()
		}
		var ePoint = theEvent.locationInWindow
		while loop {
			guard let nEvent = window?.nextEvent(matching: [.leftMouseUp, .leftMouseDragged]) else {
				continue
			}
			var thePoint = nEvent.locationInWindow
			thePoint = convert(thePoint, from: nil)
			inside = NSMouseInRect(thePoint, bounds, isFlipped)
			switch nEvent.type {
			case .leftMouseUp:
				loop = false
				if !wasdragged, wasSelected {
					sender.unhilight()
					if let idx = selectedList.firstIndex(of: sender) {
						selectedList.remove(at: idx)
					}
				} else {
					for theTad in selectedList {
						theTad.setFromPosWith(beatScale: beatScale, frequencyScale: frequencyScale)
						theTad.isMoving = false
					}
				}
				
			case .leftMouseDragged:
				wasdragged = true
				for theTad in selectedList {
					theTad.isMoving = true
					theTad.setFrameOrigin(NSPoint(x: (theTad.frame.origin.x + thePoint.x - ePoint.x), y: (theTad.frame.origin.y + thePoint.y - ePoint.y)))
				}
				ePoint = thePoint
				
			default:
				// Makes explicit to the compiler we do nothing with other events.
				break;
			}
		}
	}
}
